src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/Everything/PatientEverythingService.cs

```diff
@@ -0,0 +1,369 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Core.Features.Definition;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Features.Search;
+using Microsoft.Health.Fhir.Core.Models;
+using CompartmentType = Microsoft.Health.Fhir.ValueSets.CompartmentType;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Everything
+{
+    public class PatientEverythingService : IPatientEverythingService
+    {
+        private readonly Func<IScoped<ISearchService>> _searchServiceFactory;
+        private readonly ISearchOptionsFactory _searchOptionsFactory;
+        private readonly ISearchParameterDefinitionManager _searchParameterDefinitionManager;
+        private readonly ICompartmentDefinitionManager _compartmentDefinitionManager;
+
+        private IReadOnlyList<string> _includes = new[] { "general-practitioner", "organization" };
+        private readonly Tuple<string, string> _revinclude = Tuple.Create("Device", "patient");
+
+        public PatientEverythingService(
+            Func<IScoped<ISearchService>> searchServiceFactory,
+            ISearchOptionsFactory searchOptionsFactory,
+            ISearchParameterDefinitionManager searchParameterDefinitionManager,
+            ICompartmentDefinitionManager compartmentDefinitionManager)
+        {
+            EnsureArg.IsNotNull(searchServiceFactory, nameof(searchServiceFactory));
+            EnsureArg.IsNotNull(searchOptionsFactory, nameof(searchOptionsFactory));
+            EnsureArg.IsNotNull(searchParameterDefinitionManager, nameof(searchParameterDefinitionManager));
+            EnsureArg.IsNotNull(compartmentDefinitionManager, nameof(compartmentDefinitionManager));
+
+            _searchServiceFactory = searchServiceFactory;
+            _searchOptionsFactory = searchOptionsFactory;
+            _searchParameterDefinitionManager = searchParameterDefinitionManager;
+            _compartmentDefinitionManager = compartmentDefinitionManager;
+        }
+
+        public async Task<SearchResult> SearchAsync(
+            string resourceType,
+            string resourceId,
+            PartialDateTime start,
+            PartialDateTime end,
+            PartialDateTime since,
+            string type,
+            int? count,
```

Comment:
>int? count, [](http://example.com/codeflow?start=12&length=11)

I don't think it's been validated and used anywhere.

Regarding validation, it probably should be some reasonalbe value, from 1. to some const in service or config value.

Reply:
Yes, `count` is not used anywhere. In the [spec](https://www.hl7.org/fhir/patient-operation-everything.html) of Patient $everything, there is a `_count` input parameter. Currently it is not supported, as we use `_include` to pull the patient's direct references and resources pulled by `_include` are not limited by `count`. We may support it in the future so I just put a variable in the interface.

Reply:
searchOptions has field `IncludeCount` which I believe should control amount of data returned in include.

>Alternatively, clients may choose to page through the result set (or servers may require this). Paging through the results is done the same as for Searching, using the _count parameter, and Bundle links.

I think spec is quite clear about count use. If we have 1000 results in database, I should be able to pass count=100 and iterate over that collection with continuation token. 

Note that count is upper limit to return results, it's ok to return 10 results if count set to be 20, as long as we fetch all data via continuation token, but in same time it's not ok to return 21 result. (Fine to underdeliver but never ok to overdeliver)

Reply:
Yes, searchOptions has a field `IncludeCount` to limit the number of resources searched by `_include`. But in the scenario of Patient $everything I think there are two things that worth attention.

1). There's no continuation token provided for resources searched by `_include`. In Cosmos DB, the `_include` IDs are pulled together with the Patient resource. Then `_include` resources themselves are pulled by `_fhirDataStore.GetAsync()`. They are pulled from database at once without a continuation token, which is very different from other search queries and this would require recording extra information to support `_count`.

Taking following case as an example, if there are 10 `_include` resources, when the user specifies `_count` to 3, we will need to record extra information to pull the next 3 resources in subsequent pages (like record current returned number in continuation token, perform the same `_include` search again and return the next batch of resources).

2). I agree that from the implementation level, we could possibly figure out some way to achieve `_count`, but I also think that this requires more tests to make sure it works smoothly, because one different point between Patient $everything and other searches is that it could possibly issue more queries into the database to construct one page (searching Patient, `_include`, compartment with data, compartment without data, Device).

Given that we need to support Patient $everything for CMS requirement before July 1st and `_count` is considered low priority in our PM spec, we have come to an agreement to skip supporting `_count` in the initial release. We can try to support this in future releases after more detailed design and thorough tests.

Reply:
I can be wrong, but in my head `_count` and `continuationToken` is pair which goes together and only together. If `_count` is not supported we need to return all results in bundle. If `_count` supported we need to iterate over collection with support of `_continuationToken`.

If `_count` is low priority, I would suggest to just exclude it from all interfaces together with `_continuationToken`

Reply:
In the [spec](https://www.hl7.org/fhir/http.html#paging) of paging, there's no requirement about supporting `_count` or making it together with the continuation token, so I think it would be OK to skip `_count` here. Besides, if we do not have a continuation token, we will need to return all resources (including compartment resources) in one response, which can too large to work.